Improve observation by providing more tracing tags

cmergenthaler opened the pull request:
Opentelemetry defines standards for specifying and providing proper information for tracing spans (so called tags). 
This PR improves the observation by providing proper opentelemetry defined tags.

Fixes #2609 

As discussed in the issue, the remote service name is not explicitly set, because optentelemetry defines it as the `name of the broker or service the message will be sent to`, which would be a list of multiple brokers in the kafka service. For that reason, users are encouraged to override `KafkaTemplateObservationConvention` and `KafkaListenerObservationConvention` if they want it to have a unique name instead of the cluster-id.

_Note: Some informations are not available on the time the observation in the `KafkaTemplate` starts and would be null. Therefore I have not included in this PR, if you have suggestions on that feel free to share._

pivotal-cla commented:
@cmergenthaler Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=2616)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=2616) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@cmergenthaler Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=2616)!

garyrussell commented:
Please also fix the checkstyle problems; run `./gradlew clean check`.

garyrussell commented on spring-kafka/src/test/java/org/springframework/kafka/support/micrometer/ObservationTests.java:
```diff
@@ -78,19 +78,18 @@
+				  @Autowired SimpleTracer tracer, @Autowired KafkaListenerEndpointRegistry rler,
+				  @Autowired MeterRegistry meterRegistry, @Autowired EmbeddedKafkaBroker broker,
+				  @Autowired KafkaListenerEndpointRegistry endpointRegistry, @Autowired KafkaAdmin admin)
+			throws InterruptedException, ExecutionException, TimeoutException {
```
No leading spaces allowed.

garyrussell reviewed (changes requested):
Just minor formatting and the send Vs. publish issue.

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaTemplateObservation.java:
```diff
@@ -90,13 +138,17 @@ public static class DefaultKafkaTemplateObservationConvention implements KafkaTe
 		@Override
 		public String getContextualName(KafkaRecordSenderContext context) {
-			return context.getDestination() + " send";
+			return context.getDestination() + " publish";
```
I think, Gary, when they asked us for this, the OTEL really promoted a `send`.
The `publish` has come later on.

I might be mistaken. Let's double check with @marcingrzejszczak!

garyrussell replied:
Even if we change it; it has to be an option; I don't think we can make a change like this in a patch release.

garyrussell replied:
Presumably, we are going to need something similar for RabbitMQ too.

artembilan replied:
Well, I think that `publish`/`send` is still just a recommendation.

I suggest to stick with a `send` so far until the next `3.1` this Fall.
If other change are additive and don't override anything existing, we can apply them with this PR.

marcingrzejszczak replied:
We never said that we are going to be OpenTelemetry compliant. OTel has a tendency to do breaking changes in semantic conventions in minor / patch releases. This is not the way that we would like to go with. At the time of doing this instrumentation AFAIR the OTel recommendation was `send` and I think we should keep it that way. One can use a `ObservationFilter` to change the name from `send` to `publish` if they want to be compliant with the current version of OTel semantic conventions.

artembilan replied:
Thank you, Marcin!

So, our conclusion is like this then: we accept this change, but not with a `publish` move. We stick with `send` for time being.

I think such a statement about an OTel compliant must be mentioned somewhere in the Micrometer Tracing docs: https://micrometer.io/docs/tracing#_micrometer_tracing_opentelemetry_setup.

marcingrzejszczak replied:
> So, our conclusion is like this then: we accept this change, but not with a publish move. We stick with send for time being.

Correct!

> I think such a statement about an OTel compliant must be mentioned somewhere in the Micrometer Tracing docs: https://micrometer.io/docs/tracing#_micrometer_tracing_opentelemetry_setup.

We were looking at the OTel semantic conventions at the point of time where things got instrumented but we were never saying that we want to be OTel compliant because of the reasons I mentioned above. We never explicitly stated that we are OTel compliant really so I don't think we should tell people about it.

Also naming is unrelated to the tracer library used under the hood. So naming can be compliant with OTel semantic conventions even if you're using Brave.

artembilan replied:
Sure! No problem!
But why then people push for Otel everywhere? 😄 
It feels like they took an idea for all of these tags to be in line with Otel from somewhere on top.

See the heading of this PR, for example:

> Opentelemetry defines standards for specifying and providing proper information for tracing spans

If we don't say anywhere in our docs that we follow so-called standards then how people came to such a conclusion?

cmergenthaler commented:
Thanks for the reviews, @garyrussell & @artembilan!
I have updated the files as you said. I also moved partition and offset to high-cardinality keys as the values change a lot and while load testing my service we have encountered memory issues when having them as low-cardinality (metrics also get them as tags).

I think having different implementations that the user can use would be very helpful - maybe on all related spring projects (e.g. spring-data-mongo).

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaRecordSenderContext.java:
```diff
@@ -56,5 +57,4 @@ public String getBeanName() {
 	public String getDestination() {
 		return this.destination;
 	}
-
```
Blank lines between methods.
And therefore no changes in this class at all.
So, please, revert it to the `main` version.

artembilan reviewed (changes requested):
I'm OK with this so far.

One change request and also see a failing test:
```
ObservationTests > endToEnd(Listener, KafkaTemplate, SimpleTracer, KafkaListenerEndpointRegistry, MeterRegistry, EmbeddedKafkaBroker, KafkaListenerEndpointRegistry, KafkaAdmin, KafkaTemplate, Config) FAILED
    java.lang.AssertionError: 
    Expecting map:
      {"bar"="some bar value", "foo"="some foo value", "messaging.consumer.id"="obs1 - consumer-obs1-333", "messaging.kafka.client_id"="consumer-obs1-333", "messaging.kafka.consumer.group"="obs1", "messaging.kafka.message.offset"="0", "messaging.kafka.source.partition"="0", "messaging.operation"="receive", "messaging.source.kind"="topic", "messaging.source.name"="observation.testT1", "messaging.system"="kafka", "spring.kafka.listener.id"="obs1-0"}
    to contain entries:
      ["messaging.source.name"="observation.testT1",
        "messaging.consumer.id"="obs1 - consumer-obs1-3",
        "messaging.kafka.source.partition"="0",
        "messaging.kafka.message.offset"="0",
        "messaging.kafka.client_id"="consumer-obs1-3",
        "messaging.operation"="receive",
        "messaging.kafka.consumer.group"="obs1",
        "bar"="some bar value",
        "messaging.source.kind"="topic",
        "foo"="some foo value",
        "messaging.system"="kafka",
        "spring.kafka.listener.id"="obs1-0"]
    but the following map entries had different values:
      ["messaging.consumer.id"="obs1 - consumer-obs1-333" (expected: "obs1 - consumer-obs1-3"), "messaging.kafka.client_id"="consumer-obs1-333" (expected: "consumer-obs1-3")]
        at org.springframework.kafka.support.micrometer.ObservationTests.endToEnd(ObservationTests.java:119)
```

@garyrussell ,

do you still insist that we don't expose those new tags by default, but rather provide other `Convention` impl which would give us that info?
Although the change in the listener container and Kafka template must be there to be able to expose that info with any custom convention...

garyrussell commented:
I just took another look and some of the new tags are high cardinality (..._PARTITION, ..._OFFSET) but they are in the `ListenerLowCardinalityTags` enum.

Also, with high concurrency, the client id can be high cardinality too; each child container gets its own client id.

My preference is to make this an alternative observation, even when these cardinality issues are fixed. But, yes, the code to create the context will have to remain.

Wzy19930507 commented:
@cmergenthaler Hello, if you don't mind, i want work on this PR, please give me Developer role to push code, thanks.

artembilan commented:
@Wzy19930507 ,

thank you for considering!

This PR is one year old, so might be better to start from scratch, even if you will use same code base from this PR and mention original author.

artembilan commented:
Superseded by #3169
